File: ocaml/test/test_json_output.ml
let () =
  let open Alcotest in
  let open Ccc_lib in
  run "json_output" [
    ("parse_opencode", [
      test_case "response" `Quick (fun () ->
        let r = Json_output.parse_opencode_json {|{"response": "hello"}
|} in
        check int "events" 1 (List.length r.Json_output.events);
        check string "type" "text" (List.hd r.Json_output.events).Json_output.je_event_type;
        check string "text" "hello" (List.hd r.Json_output.events).Json_output.je_text;
        check string "final" "hello" r.Json_output.final_text);

      test_case "error" `Quick (fun () ->
        let r = Json_output.parse_opencode_json {|{"error": "fail"}
|} in
        check string "error" "fail" r.Json_output.error_text;
        check string "type" "error" (List.hd r.Json_output.events).Json_output.je_event_type);

      test_case "skips invalid json" `Quick (fun () ->
        let r = Json_output.parse_opencode_json "bad\n{\"response\": \"ok\"}\n" in
        check int "events" 1 (List.length r.Json_output.events);
        check string "final" "ok" r.Json_output.final_text);

      test_case "empty input" `Quick (fun () ->
        let r = Json_output.parse_opencode_json "" in
        check int "events" 0 (List.length r.Json_output.events));
    ]);

    ("parse_claude_code", [
      test_case "system init" `Quick (fun () ->
        let r = Json_output.parse_claude_code_json {|{"type":"system","subtype":"init","session_id":"s1"}
|} in
        check string "session" "s1" r.Json_output.session_id);

      test_case "assistant" `Quick (fun () ->
        let r = Json_output.parse_claude_code_json {|{"type":"assistant","message":{"content":[{"type":"text","text":"hi"}]}}
|} in
        check int "events" 1 (List.length r.Json_output.events);
        check string "type" "assistant" (List.hd r.Json_output.events).Json_output.je_event_type;
        check string "final" "hi" r.Json_output.final_text);

      test_case "text delta" `Quick (fun () ->
        let r = Json_output.parse_claude_code_json {|{"type":"stream_event","event":{"type":"content_block_delta","delta":{"type":"text_delta","text":"yo"}}}}
|} in
        check string "type" "text_delta" (List.hd r.Json_output.events).Json_output.je_event_type;
        check string "text" "yo" (List.hd r.Json_output.events).Json_output.je_text);

      test_case "tool use" `Quick (fun () ->
        let r = Json_output.parse_claude_code_json {|{"type":"tool_use","tool_name":"read","tool_input":{"a":1}}
|} in
        let ev = List.hd r.Json_output.events in
        check string "type" "tool_use" ev.Json_output.je_event_type;
        check string "name" "read" (Option.get ev.Json_output.je_tool_call).Json_output.tc_name);

      test_case "tool result" `Quick (fun () ->
        let r = Json_output.parse_claude_code_json {|{"type":"tool_result","tool_use_id":"t1","content":"out","is_error":false}}
|} in
        let tr = Option.get (List.hd r.Json_output.events).Json_output.je_tool_result in
        check string "id" "t1" tr.Json_output.tr_tool_call_id;
        check string "content" "out" tr.Json_output.tr_content;
        check bool "is_error" false tr.Json_output.tr_is_error);

      test_case "result success" `Quick (fun () ->
        let r = Json_output.parse_claude_code_json {|{"type":"result","subtype":"success","result":"done","cost_usd":0.1,"duration_ms":500}}
|} in
        check string "final" "done" r.Json_output.final_text;
        check (float 0.001) "cost" 0.1 r.Json_output.cost_usd;
        check int "duration" 500 r.Json_output.duration_ms);

      test_case "result error" `Quick (fun () ->
        let r = Json_output.parse_claude_code_json {|{"type":"result","subtype":"error","error":"boom"}}
|} in
        check string "error" "boom" r.Json_output.error_text);
    ]);

    ("parse_kimi", [
      test_case "assistant text" `Quick (fun () ->
        let r = Json_output.parse_kimi_json {|{"role":"assistant","content":"hello"}
|} in
        check string "type" "assistant" (List.hd r.Json_output.events).Json_output.je_event_type;
        check string "final" "hello" r.Json_output.final_text);

      test_case "tool calls" `Quick (fun () ->
        let r = Json_output.parse_kimi_json {|{"role":"assistant","content":"","tool_calls":[{"id":"1","function":{"name":"bash","arguments":"{}"}}]}}
|} in
        let tc_ev = List.find (fun e -> e.Json_output.je_event_type = "tool_call") r.Json_output.events in
        check string "name" "bash" (Option.get tc_ev.Json_output.je_tool_call).Json_output.tc_name);

      test_case "passthrough" `Quick (fun () ->
        let r = Json_output.parse_kimi_json {|{"type":"TurnBegin"}}
|} in
        check string "type" "turnbegin" (List.hd r.Json_output.events).Json_output.je_event_type);
    ]);

    ("parse_json_output", [
      test_case "dispatches opencode" `Quick (fun () ->
        let r = Json_output.parse_json_output {|{"response":"ok"}
|} "opencode" in
        check string "schema" "opencode" r.Json_output.schema_name;
        check string "final" "ok" r.Json_output.final_text);

      test_case "unknown schema" `Quick (fun () ->
        let r = Json_output.parse_json_output "" "unknown" in
        check bool "has error" true (String.length r.Json_output.error_text > 0));
    ]);

    ("render_parsed", [
      test_case "text events" `Quick (fun () ->
        let r = { Json_output.schema_name = "test"; Json_output.events = [
          { Json_output.je_event_type = "text"; je_text = "hello"; je_thinking = ""; je_tool_call = None; je_tool_result = None };
          { Json_output.je_event_type = "assistant"; je_text = "world"; je_thinking = ""; je_tool_call = None; je_tool_result = None };
        ]; Json_output.final_text = ""; Json_output.session_id = ""; Json_output.error_text = "";
          Json_output.cost_usd = 0.0; Json_output.duration_ms = 0 } in
        check string "render" "hello\nworld" (Json_output.render_parsed r));

      test_case "thinking" `Quick (fun () ->
        let r = { Json_output.schema_name = "test"; Json_output.events = [
          { Json_output.je_event_type = "thinking"; je_text = ""; je_thinking = "hmm"; je_tool_call = None; je_tool_result = None };
        ]; Json_output.final_text = ""; Json_output.session_id = ""; Json_output.error_text = "";
          Json_output.cost_usd = 0.0; Json_output.duration_ms = 0 } in
        check string "render" "[thinking] hmm" (Json_output.render_parsed r));

      test_case "tool use and result" `Quick (fun () ->
        let r = { Json_output.schema_name = "test"; Json_output.events = [
          { Json_output.je_event_type = "tool_use"; je_text = ""; je_thinking = "";
            je_tool_call = Some { Json_output.tc_id = ""; tc_name = "read"; tc_arguments = "" };
            je_tool_result = None };
          { Json_output.je_event_type = "tool_result"; je_text = ""; je_thinking = "";
            je_tool_call = None;
            je_tool_result = Some { Json_output.tr_tool_call_id = ""; tr_content = "output"; tr_is_error = false } };
        ]; Json_output.final_text = ""; Json_output.session_id = ""; Json_output.error_text = "";
          Json_output.cost_usd = 0.0; Json_output.duration_ms = 0 } in
        check string "render" "[tool] read\n[tool_result] output" (Json_output.render_parsed r));

      test_case "fallback" `Quick (fun () ->
        let r = { Json_output.schema_name = "test"; Json_output.events = [];
          Json_output.final_text = "fallback"; Json_output.session_id = ""; Json_output.error_text = "";
          Json_output.cost_usd = 0.0; Json_output.duration_ms = 0 } in
        check string "render" "fallback" (Json_output.render_parsed r));

      test_case "error" `Quick (fun () ->
        let r = { Json_output.schema_name = "test"; Json_output.events = [
          { Json_output.je_event_type = "error"; je_text = "oops"; je_thinking = ""; je_tool_call = None; je_tool_result = None };
        ]; Json_output.final_text = ""; Json_output.session_id = ""; Json_output.error_text = "";
          Json_output.cost_usd = 0.0; Json_output.duration_ms = 0 } in
        check string "render" "[error] oops" (Json_output.render_parsed r));
    ]);
  ]

File: ruby/test/test_json_output.rb
# frozen_string_literal: true

require "minitest/autorun"
require "call_coding_clis/json_output"
require "pathname"

OPENCODE_STDOUT = '{"response":"Hello from OpenCode"}'

CLAUDE_CODE_STDOUT = <<~LINES
  {"type":"system","subtype":"init","session_id":"sess-1","tools":[],"model":"mock","permission_mode":"default"}
  {"type":"assistant","message":{"id":"msg_1","type":"message","role":"assistant","model":"mock","content":[{"type":"text","text":"I will help you."}],"stop_reason":"end_turn","stop_sequence":null,"usage":{"input_tokens":10,"output_tokens":5}},"session_id":"sess-1"}
  {"type":"result","subtype":"success","cost_usd":0.003,"duration_ms":100,"duration_api_ms":80,"num_turns":1,"result":"I will help you.","session_id":"sess-1","usage":{"input_tokens":10,"output_tokens":5}}
LINES

CLAUDE_CODE_TOOL_USE = <<~LINES
  {"type":"system","subtype":"init","session_id":"sess-2","tools":[],"model":"mock","permission_mode":"default"}
  {"type":"tool_use","tool_name":"read_file","tool_input":{"path":"/test.py"},"session_id":"sess-2"}
  {"type":"tool_result","tool_use_id":"toolu_1","content":"file contents","is_error":false,"session_id":"sess-2"}
  {"type":"result","subtype":"success","result":"Done","session_id":"sess-2","cost_usd":0.0,"duration_ms":200,"duration_api_ms":150,"num_turns":2,"usage":{"input_tokens":20,"output_tokens":10}}
LINES

KIMI_STDOUT = '{"role":"assistant","content":"Hello from Kimi"}'

KIMI_WITH_TOOL_CALLS = <<~LINES
  {"role":"assistant","content":"Let me read that.","tool_calls":[{"type":"function","id":"call_1","function":{"name":"read_file","arguments":"{\\"path\\":\\"/src/main.py\\"}"}}]}
  {"role":"tool","content":[{"type":"text","text":"<system>Tool completed successfully.</system>"},{"type":"text","text":"file contents"}],"tool_call_id":"call_1"}
  {"role":"assistant","content":"Done reading the file."}
LINES

KIMI_THINKING = <<~LINES
  {"role":"assistant","content":[{"type":"think","think":"I need to analyze this...","encrypted":null},{"type":"text","text":"Here is my answer"}]}
LINES

FIXTURES = Pathname.new(__dir__).join("..", "..", "tests", "fixtures", "runner-transcripts")

class TestParseOpenCodeJson < Minitest::Test
  def test_simple_response
    result = CallCodingClis::JsonOutput.parse_json_output(OPENCODE_STDOUT, "opencode")
    assert_equal "opencode", result.schema_name
    assert_equal "Hello from OpenCode", result.final_text
    assert_equal 1, result.events.length
    assert_equal "text", result.events[0].event_type
  end

  def test_error_response
    stdout = '{"error":"something went wrong"}'
    result = CallCodingClis::JsonOutput.parse_opencode_json(stdout)
    assert_equal "something went wrong", result.error
    assert_equal 1, result.events.length
    assert_equal "error", result.events[0].event_type
  end
end

class TestParseClaudeCodeJson < Minitest::Test
  def test_simple_response
    result = CallCodingClis::JsonOutput.parse_json_output(CLAUDE_CODE_STDOUT, "claude-code")
    assert_equal "claude-code", result.schema_name
    assert_equal "I will help you.", result.final_text
    assert_equal "sess-1", result.session_id
    assert_in_delta 0.003, result.cost_usd
    assert_equal 100, result.duration_ms
  end

  def test_tool_use_and_result
    result = CallCodingClis::JsonOutput.parse_claude_code_json(CLAUDE_CODE_TOOL_USE)
    assert_equal "Done", result.final_text
    tool_events = result.events.select { |e| e.event_type == "tool_use" }
    assert_equal 1, tool_events.length
    refute_nil tool_events[0].tool_call
    assert_equal "read_file", tool_events[0].tool_call.name
  end

  def test_tool_result
    result = CallCodingClis::JsonOutput.parse_claude_code_json(CLAUDE_CODE_TOOL_USE)
    tr_events = result.events.select { |e| e.event_type == "tool_result" }
    assert_equal 1, tr_events.length
    tr = tr_events[0].tool_result
    refute_nil tr
    assert_equal "toolu_1", tr.tool_call_id
    assert_equal "file contents", tr.content
    refute tr.is_error
  end

  def test_error_result
    stdout = '{"type":"result","subtype":"error","error":"rate limited","session_id":"s1"}'
    result = CallCodingClis::JsonOutput.parse_claude_code_json(stdout)
    assert_equal "rate limited", result.error
  end
end

class TestParseKimiJson < Minitest::Test
  def test_simple_response
    result = CallCodingClis::JsonOutput.parse_json_output(KIMI_STDOUT, "kimi")
    assert_equal "kimi", result.schema_name
    assert_equal "Hello from Kimi", result.final_text
  end

  def test_tool_calls
    result = CallCodingClis::JsonOutput.parse_kimi_json(KIMI_WITH_TOOL_CALLS)
    assert_equal "Done reading the file.", result.final_text
    tc_events = result.events.select { |e| e.event_type == "tool_call" }
    assert_equal 1, tc_events.length
    tc = tc_events[0].tool_call
    refute_nil tc
    assert_equal "read_file", tc.name
  end

  def test_tool_result
    result = CallCodingClis::JsonOutput.parse_kimi_json(KIMI_WITH_TOOL_CALLS)
    tr_events = result.events.select { |e| e.event_type == "tool_result" }
    assert_equal 1, tr_events.length
    tr = tr_events[0].tool_result
    refute_nil tr
    assert_equal "file contents", tr.content
  end

  def test_thinking
    result = CallCodingClis::JsonOutput.parse_kimi_json(KIMI_THINKING)
    think_events = result.events.select { |e| e.event_type == "thinking" }
    assert_equal 1, think_events.length
    assert_equal "I need to analyze this...", think_events[0].thinking
    assert_equal "Here is my answer", result.final_text
  end
end

class TestRenderParsed < Minitest::Test
  def test_render_opencode
    result = CallCodingClis::JsonOutput.parse_opencode_json(OPENCODE_STDOUT)
    rendered = CallCodingClis::JsonOutput.render_parsed(result)
    assert_equal "Hello from OpenCode", rendered
  end

  def test_render_claude
    result = CallCodingClis::JsonOutput.parse_claude_code_json(CLAUDE_CODE_STDOUT)
    rendered = CallCodingClis::JsonOutput.render_parsed(result)
    assert_includes rendered, "I will help you."
  end

  def test_render_kimi
    result = CallCodingClis::JsonOutput.parse_kimi_json(KIMI_STDOUT)
    rendered = CallCodingClis::JsonOutput.render_parsed(result)
    assert_equal "Hello from Kimi", rendered
  end

  def test_render_with_thinking
    result = CallCodingClis::JsonOutput.parse_kimi_json(KIMI_THINKING)
    rendered = CallCodingClis::JsonOutput.render_parsed(result)
    assert_includes rendered, "[thinking]"
    assert_includes rendered, "Here is my answer"
  end

  def test_render_with_tool_use
    result = CallCodingClis::JsonOutput.parse_claude_code_json(CLAUDE_CODE_TOOL_USE)
    rendered = CallCodingClis::JsonOutput.render_parsed(result)
    assert_includes rendered, "[tool] read_file"
  end

  def test_render_unknown_schema
    result = CallCodingClis::JsonOutput.parse_json_output("{}", "unknown-schema")
    assert_equal "unknown schema: unknown-schema", result.error
  end

  def test_empty_input
    result = CallCodingClis::JsonOutput.parse_opencode_json("")
    assert_equal "", result.final_text
  end

  def test_malformed_json_skipped
    stdout = "not json\n#{OPENCODE_STDOUT}"
    result = CallCodingClis::JsonOutput.parse_opencode_json(stdout)
    assert_equal "Hello from OpenCode", result.final_text
  end

  def test_fixture_claude_stream_unknown_events
    raw = FIXTURES.join("claude", "stream_unknown_events", "stdout.ndjson").read
    result = CallCodingClis::JsonOutput.parse_claude_code_json(raw)
    rendered = CallCodingClis::JsonOutput.render_parsed(result)
    assert_includes rendered, "Computing the first multiplication."
    assert_operator result.raw_lines.length, :>=, 10
    assert result.raw_lines.any? { |obj| obj["type"] == "rate_limit_event" }
    assert result.raw_lines.any? { |obj| obj["type"] == "stream_event" && obj.dig("event", "type") == "message_start" }
    assert result.raw_lines.any? { |obj| obj["type"] == "stream_event" && obj.dig("event", "type") == "message_delta" }
    assert result.raw_lines.any? { |obj| obj["type"] == "stream_event" && obj.dig("event", "type") == "message_stop" }
    assert result.raw_lines.any? { |obj| obj["type"] == "stream_event" && obj.dig("event", "delta", "type") == "signature_delta" }
  end
end
